src/coreclr/jit/importercalls.cpp

```diff
@@ -4169,6 +4154,328 @@ GenTree* Compiler::impSRCSUnsafeIntrinsic(NamedIntrinsic        intrinsic,
+                // * 64-bit lzcnt: (value == 0) ? 64 : (63 ^ BSR(value))
+
+                GenTree* op1Dup;
+                GenTree* op1 = impCloneExpr(impPopStack().val, &op1Dup, NO_CLASS_HANDLE, (unsigned)CHECK_SPILL_ALL,
+                                            nullptr DEBUGARG("Cloning op1 for LeadingZeroCount"));
+
+                hwintrinsic = varTypeIsLong(retType) ? NI_X86Base_X64_BitScanReverse : NI_X86Base_BitScanReverse;
+                op1Dup      = gtNewScalarHWIntrinsicNode(retType, op1, hwintrinsic);
+
+                GenTree*      cond     = gtNewOperNode(GT_EQ, retType, op1, gtNewIconNode(0, retType));
+                GenTree*      trueRes  = gtNewIconNode(varTypeIsLong(retType) ? 64 : 32, retType);
+                GenTree*      icon     = gtNewIconNode(varTypeIsLong(retType) ? 63 : 31, retType);
+                GenTree*      falseRes = gtNewOperNode(GT_XOR, retType, op1Dup, icon);
+                GenTreeColon* colon    = gtNewColonNode(retType, trueRes, falseRes);
+
+                return gtNewQmarkNode(retType, cond, colon);
+            }
+#elif defined(TARGET_ARM64)
+            // Pop the value from the stack
+            impPopStack();
+
+            hwintrinsic = varTypeIsLong(retType) ? NI_ArmBase_Arm64_LeadingZeroCount : NI_ArmBase_LeadingZeroCount;
+            return gtNewScalarHWIntrinsicNode(retType, op1, hwintrinsic);
+#endif // TARGET_*
+#endif // FEATURE_HW_INTRINSICS
+
+            return nullptr;
+        }
+
+        case NI_PRIMITIVE_Log2:
+        {
+            assert(sig->numArgs == 1);
+
+            var_types baseType = JitType2PreciseVarType(sig->retType);
+            assert(!varTypeIsSmall(baseType));
+
+            if (op1->IsIntegralConst())
+            {
+                if (varTypeIsLong(retType))
+                {
+                    uint64_t cns = static_cast<uint64_t>(op1->AsIntConCommon()->LngValue());
+
+                    if (varTypeIsSigned(baseType) && (static_cast<int64_t>(cns) < 0))
+                    {
+                        // Throwing case, fallback to software impl
+                        return nullptr;
+                    }
+
+                    return gtNewIconNode(BitOperations::Log2(cns), retType);
+                }
+                else
+                {
+                    uint32_t cns = static_cast<uint32_t>(op1->AsIntConCommon()->IconValue());
+
+                    if (varTypeIsSigned(baseType) && (static_cast<int64_t>(cns) < 0))
+                    {
+                        // Throwing case, fallback to software impl
+                        return nullptr;
+                    }
+
+                    return gtNewIconNode(BitOperations::Log2(cns), retType);
+                }
+            }
+
+            if (varTypeIsSigned(baseType))
+            {
+                // Signed needs to throw for negative inputs, so fallback to software impl
+                return nullptr;
```

Comment:
We could handle this specially with a check for negative values, but it is a more complex change and so I decided to push it out to a later PR.